plugin-management-library/src/test/java/io/jenkins/tools/pluginmanager/impl/PluginManagerIntegrationTest.java

```diff
@@ -0,0 +1,93 @@
+package io.jenkins.tools.pluginmanager.impl;
+
+import hudson.util.VersionNumber;
+import io.jenkins.tools.pluginmanager.config.Config;
+import io.jenkins.tools.pluginmanager.config.Settings;
+import org.apache.commons.io.IOUtils;
+import org.json.JSONObject;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.contrib.java.lang.system.SystemOutRule;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.Arrays;
+
+import static org.hamcrest.CoreMatchers.*;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for {@link PluginManager} which operate with real data.
+ * No mocks here.
+ */
+public class PluginManagerIntegrationTest {
+
+    static JSONObject latestUcJson;
+
+    @ClassRule
+    public static TemporaryFolder tmp = new TemporaryFolder();
+    public static File jenkinsWar;
+    public static File cacheDir;
+    public static File pluginsDir;
+
+    @Rule
+    public final SystemOutRule systemOutRule = new SystemOutRule().enableLog();
+
+    public interface Configurator {
+        void configure(Config.Builder configBuilder);
+    }
+
+    public static PluginManager initPluginManager(Configurator configurator) throws IOException {
+        Config.Builder configBuilder = Config.builder()
+                .withJenkinsWar(jenkinsWar.getAbsolutePath())
+                .withPluginDir(pluginsDir)
+                .withShowAvailableUpdates(true)
+                .withIsVerbose(true)
+                .withDoDownload(false);
+        configurator.configure(configBuilder);
+        Config config = configBuilder.build();
+
+        PluginManager pluginManager = new PluginManager(config);
+        pluginManager.setCm(new CacheManager(cacheDir.toPath(), true));
+        pluginManager.setJenkinsVersion(new VersionNumber("2.222.1"));
+        pluginManager.setLatestUcJson(latestUcJson);
+        pluginManager.setLatestUcPlugins(latestUcJson.getJSONObject("plugins"));
+        pluginManager.setPluginInfoJson(pluginManager.getJson(new URL(Settings.DEFAULT_PLUGIN_INFO_LOCATION), "plugin-versions"));
+
+        return pluginManager;
+    }
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        String jsonString = IOUtils.toString(PluginManagerIntegrationTest.class.getResource("updates.json"), StandardCharsets.UTF_8);
```

Comment:
Not sure it is a good idea to keep this file as resource, it is almost 2Mb.
But there is no way to create reliable tests without persisting the file somewhere.

Opinions @jenkinsci/plugin-installation-manager-tool-developers ?

Reply:
just add it, it's only 2mb